### ISSUE TITLE
Manage API keys in web interface

### ISSUE THREAD

**Issue by peter-** [filesender/filesender#579]
Filesender now provides a link to the Python API client in the user profile but does not provide any means to create or reset (i.e., revoke and create in one step) the necessary API keys required for that. (Update: If `auth_remote_user_autogenerate_secret=true` is set in the config it will create and show such API keys. AFAICT that's the only method FileSender provides atm.)

As [suggested earlier](https://sympa.uninett.no/lists/filesender.org/arcsearch_id/filesender-dev/2018-03/20180309160039.3yhadbfovt6cwsor%40aco.net) modelling this after Gitlab's "private access tokens" (i.e., not OAuth) would be sufficient, IMO: A bit of UI to "create/reset API key" (in the user profile, right next to the Python client) and code to persist that key (salted and hashed, like a password) to the database, not to be displayed again after creation. (You'd reset/regenerate it, instead, had you forgotten/lost the API key.)

In the initial implementaton a single API could be managed that way (K.I.S.S.). A more flexible implementation might allow to provide "labels" for those keys when creating them and manage multiple API keys, so the subject knows what API key was used where (if the label them accordingly). That avoids having to share the same API across multiple machines and then possibly having to re-key all clients in case *one* of those machines might not be considered sufficiently secure. (But whether people actually make use of the CLI client and from multiple machines is an open question to me, so the simple case should probably be looked at first.)

**Comment by monkeyiq** [comment]
I just merged https://github.com/filesender/filesender/pull/718 which allows the api secret to be deleted and created. The time of creation is now also stored and displayed. There is also support for an aup for the secret if a site wishes to enable that. Though I will mention the aup part elsewhere.

I am not sure what extra security you would get by not displaying the api secret after the initial creation. As soon as a user discovered that the site was doing that they would likely copy and paste it into a text file to mitigate the site hiding it.

There is still only a single api secret per user. Perhaps if we are planning to move to many secrets it might be a good time to first move to oauth or something like that. Another interesting idea might be to allow server side checks on api secret use. For example, a user might know that it is only used from ip address xyz and if requests come from elsewhere might like to know about it or have the key invalidated. 

In the recent updates the api secret is also downloadable in an ini file which includes other information to such as base_url. There are a number of security issues with passing the api secret on the command line as was the only option before.

**Comment by WebSpider** [comment]
@peter- has this been implemented sufficiently?

**Comment by peter-** [comment]
@WebSpider: yes, thanks. tbh I don't know why this wan't closed after #718 was merged.

Outstanding replies to @monkeyiq, for posterity:

> I am not sure what extra security you would get by not displaying the api secret after the initial creation. As soon as a user discovered that the site was doing that they would likely copy and paste it into a text file to mitigate the site hiding it.

For the exact same reason your own password will never be shown to you in any sane system. (Which doesn't remove your ability to write down copies of your password anywhere you want, securely or inscurely).  And of course people *will* be writing their API secret into their `.filesender.ini` config file (and/or password managers) and so they will not need to rely on the credential to be shown in their Filesender profile page in the clear for all eternity.  
I.e., I think it does not *need* to be shown there after initial creation and therefore it probably shouldn't.  
But with #718 the desired minimal functionalty is there (to create and remove API secrets) and I'm happy with that.

> There is still only a single api secret per user.

That's also fine for an initial implementation. The only effect of not allowing more than one would be people potentially re-using their sole API secret across different systems which only has the hypothetical downside I already explained: Re-keying then always effects all systems (potentially unnecessarily).  
I personally wouldn't expect users asking for this feature since they *can* re-use a single API secret across multiple systems, thereby satisfying any use case the might have involveing the existing API client(s), even if that means a future re-keying event would cause a bit more work for them. So again, the current feature is sufficient.
